client/session.go

```diff
@@ -955,12 +978,20 @@ func (s *session) fetchAllAttempt(
 				successIters := results[:success]
 				resultsLock.RUnlock()
 				iter := s.seriesIteratorPool.Get()
-				iter.Reset(ids[idx], startInclusive, endExclusive, successIters)
+				// NB(prateek): we need to allocate a copy of tsID to allow the seriesIterator
+				// to have control over the lifecycle of ID. We cannot allow seriesIterator
+				// to control the lifecycle of the original tsID, as it might still be in use
+				// due to a pending request in queue.
+				seriesID := s.idPool.Clone(tsID)
+				iter.Reset(seriesID, startInclusive, endExclusive, successIters)
 				iters.SetAt(idx, iter)
 			}
 			if atomic.AddInt32(&resultsAccessors, -1) == 0 {
 				s.iteratorArrayPool.Put(results)
 			}
+			if atomic.AddInt32(&idAccessors, -1) == 0 {
+				tsID.Finalize()
+			}
```

Comment:
Think you need to add another block here?
```go
			if atomic.AddInt32(&namespaceAccessors, -1) == 0 {
				namespace.Finalize()
			}
```

Reply:
Nah we don't need to at this time: `allCompletionFn` doesn't have a reference to `nsID`, only `completionFn` does. Tho this will change once SeriesIterator includes a reference to the namespace it's referencing. I'll make both changes and add to this PR. Can decide once you see it.

Reply:
done